src/vs/editor/contrib/hover/browser/markdownHoverParticipant.ts

```diff
@@ -124,17 +162,162 @@ export class MarkdownHoverParticipant implements IEditorHoverParticipant<Markdow
 			return AsyncIterableObject.EMPTY;
 		}
 
-		const position = new Position(anchor.range.startLineNumber, anchor.range.startColumn);
-		return getHover(this._languageFeaturesService.hoverProvider, model, position, token)
+		this._position = anchor.range.getStartPosition();
+		return getHover(this._languageFeaturesService.hoverProvider, model, this._position, undefined, token)
 			.filter(item => !isEmptyMarkdownString(item.hover.contents))
 			.map(item => {
 				const rng = item.hover.range ? Range.lift(item.hover.range) : anchor.range;
-				return new MarkdownHover(this, rng, item.hover.contents, false, item.ordinal);
+				return new VerboseMarkdownHover(item.hover, item.provider, this, rng, false, item.ordinal);
 			});
 	}
 
-	public renderHoverParts(context: IEditorHoverRenderContext, hoverParts: MarkdownHover[]): IDisposable {
-		return renderMarkdownHovers(context, hoverParts, this._editor, this._languageService, this._openerService);
+	public renderHoverParts(context: IEditorHoverRenderContext, hoverParts: (MarkdownHover | VerboseMarkdownHover)[]): IDisposable {
+		this._context = context;
+		this._context.disposables?.add(toDisposable(() => {
+			this._hoverData.forEach(hoverData => {
+				if (hoverData instanceof VerboseMarkdownHover) {
+					hoverData.hover.dispose();
+				}
+			});
+		}));
```

Comment:
This is fishy.
When this `toDisposable` gets disposed, it will dispose the lastest `this._hoverData`!
You assign it in line 184, so it will probably be that one. But it is difficult to tell.

`this._context = context;` is also a bit fishy. Who owns `context`, i.e. who is responsibe of disposing the disposables in it?
I don't see any reason why `renderHoverParts` couldn't be called, but then the toDisposable would leak.

Reply:
Also `this._context.disposables?.add` is really not good. The disposable will leak if `_context.disposables` is undefined!!